### PR TITLE
Label propagation clustering

### PR DESCRIPTION
**Pull request by d-michail**
One of the classical clustering algorithms using "label propagation".

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/clustering/LabelPropagationClustering.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        private boolean shouldStop()
+        {
+            for (V v : graph.vertexSet()) {
+                Pair<Map<String, Integer>, Integer> labelCountsAndMaximum =
+                    getNeighborLabelCountsAndMaximum(v);
+                Map<String, Integer> counts = labelCountsAndMaximum.getFirst();
+
+                String vLabel = labels.get(v);
+                int vLabelCount = counts.getOrDefault(vLabel, 0);
+                for (E e : graph.edgesOf(v)) {
```
Comment (Toptachamann) [inline jgrapht-core/src/main/java/org/jgrapht/alg/clustering/LabelPropagationClustering.java]: As far as I understand, what we are trying to do here is to find a neighbour of `v` who's label count is higher than the label count of `v`. We can avoid traversing the neighbour edges of `v` again by traversing the `counts` map entries instead. As an alternative, it's possible to compute the desired maximum label count max value immediately while computing the `counts` map in a helper function. 

I've tested the performance difference and it was really marginal (~ 3-5% for graphs of 1000/10000 vertices generated using Kleinberg graph generator), so I'm not sure if it makes sense to apply these changes.

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/clustering/LabelPropagationClustering.java]: Yes you are right. If you have the changes somewhere, feel free to commit on this branch.

**Comment by d-michail** [comment]
I am not sure if this is required `&& !uLabel.equals(vLabel)`.

The relevant discussion in the paper is the following:

> Ideally the iterative process should continue until no node in the network changes its label. However, there could be nodes in the network that have an equal maximum number of
neighbors in two or more communities. Since we break ties randomly among the possible candidates, the labels on such nodes could change over iterations even if the labels of their
neighbors remain constant. Hence we perform the iterative process until every node in the network has a label to which the maximum number of its neighbors belongs. By doing so we obtain a partition of the network into disjoint communities, where every node has at least as many neighbors within its community as it has with any other community.

**Comment by Toptachamann** [comment]
At least the intended idea is that we compute the maximum label frequency of the neighbours of `v`, which label is not equal to the `v`'s label. 

Formally: `[label_1 -> cnt_1, ..., label_v-1 -> cnt_v-1, label_v -> cnt_v, label_v+1 -> cnt_v+1, ..., label_n -> cnt_n]` => `result = max(cnt_1, ..., cnt_v-1, cnt_v+1, ..., cnt_n)`. Now, if `result > cnt_v` => we should not stop => we should continue.

**Comment by Toptachamann** [comment]
@d-michail does it sound correct? Currently, this PR seems ready to get merged if this issue gets sorted out

**Comment by d-michail** [comment]
Yep, I took a closed look and it seems fine! Thanks!

**Comment by Toptachamann** [comment]
All right, merging this PR. Thanks!

**Pull request merged**